libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts

```diff
@@ -42,18 +43,78 @@ export interface EntityAssignment {
      * The number of times event has been raised.
      */
     raisedCount: number;
+
+    /**
+     * The expected properties when assignment was made.
+     */
+    expectedProperties: string[];
 }
 
 /**
  * Static methods for working with `EntityAssignment` objects.
  */
-export class EntityAssignment {
+export class EntityAssignment implements EntityAssignmentConfiguration {
+    /**
+     * Event name to surface
+     */
+    public event: string;
+
+    /**
+     * Name of property being assigned.
+     */
+    public property: string;
+
+    /**
+     * Operation to assign entity to property.
+     */
+    public operation: string;
+
+    /**
+     * Recognized entity value.
+     */
+    public value: Partial<EntityInfo>;
+
+    /**
+     * Alternative assignment.
+     */
+    public alternative: EntityAssignment;
+
+    /**
+     * Value indicating whether this entity was in `DialogPath.ExpectedProperties`.
+     */
+    public isExpected: boolean;
+
+    /**
+     * The number of times event has been raised.
+     */
+    public raisedCount: number;
+
+    /**
+     * The expected properties when assignment was made.
+     */
+    public expectedProperties: string[];
+
+    constructor(assignment: Partial<EntityAssignmentConfiguration>) {
+        const newEntity: Partial<EntityAssignmentConfiguration> = {
+            alternative: assignment.alternative ?? undefined,
```

Comment:
Forgot to leave that as a comment. `undefined` is necessary because there are some areas where an equality check is done on two objects and `_.isEqual()` doesn't understand that this:

```js
const obj = {
	a: 'a'
}
```

should be considered equal to:

```js
```js
const obj = {
	a: 'a',
	b: undefined
}
```

Since the constructor accepts `Partial`s, I figured the best thing to do was to just ensure all properties are added in some way or another. Very open to other ideas.

Reply:
I'm not in love with this code either, but your explanation is fair and makes sense. I would specifically cite `_.isEqual` behavior in the comment above as `undefined` and a key that doesn't exist should be semantically equivalent ([javascript.info is a good source of wisdom here](https://javascript.info/object#property-existence-test-in-operator)).

Reply:
And just to hammer it home, it's not just `lodash`. Node's `util.isDeepStrictEqual()` and `assert.deepStrictEqual()` operate the same way. I considered writing a custom equality checker, but that seemed like although equally beneficial, it was more complex. Could go either way.

Reply:
I think it's fine, the comment clearly explains what's going on. All good!